File: frontend/src/types/issues.ts
export type IssueStatus = "unresolved" | "resolved" | "ignored" | "regressed";
export type Level = "debug" | "info" | "warning" | "error" | "fatal";

export type Issue = {
  id: string;
  fingerprint: string;
  title: string;
  culprit: string | null;
  level: Level;
  status: IssueStatus;
  first_seen_at: string;
  last_seen_at: string;
  event_count: number;
  user_count: number;
};

export type IssueListResponse = {
  items: Issue[];
  next_cursor: string | null;
};

export type ProjectSummary = {
  unresolved_count: number;
  events_24h: number;
  affected_users_24h: number;
  unique_issues_24h: number;
};

export type FrequencyPoint = {
  t: string;
  count: number;
};

export type FrequencyResponse = {
  range: string;
  interval: "hour" | "day";
  points: FrequencyPoint[];
};

export type IssueListParams = {
  status?: string;
  q?: string;
  cursor?: string;
  limit?: number;
};

export type Frame = {
  function: string | null;
  module: string | null;
  filename: string | null;
  abs_path: string | null;
  lineno: number | null;
  colno: number | null;
  context_line: string | null;
  pre_context: string[];
  post_context: string[];
  in_app: boolean | null;
};

export type Stacktrace = {
  frames: Frame[];
};

export type ExceptionInfo = {
  type: string;
  value: string | null;
  module: string | null;
  stacktrace: Stacktrace | null;
};

export type BreadcrumbEntry = {
  timestamp: string | null;
  type: string | null;
  category: string | null;
  level: string | null;
  message: string | null;
  data: Record<string, unknown> | null;
};

export type EventOut = {
  id: string;
  event_id: string;
  occurred_at: string;
  received_at: string;
  environment: string;
  release: string | null;
  platform: string;
  sdk_name: string | null;
  sdk_version: string | null;
  level: Level;
  message: string | null;
  exception: ExceptionInfo | null;
  request: Record<string, unknown> | null;
  user: Record<string, unknown> | null;
  tags: Record<string, unknown>;
  contexts: Record<string, unknown>;
  breadcrumbs: BreadcrumbEntry[];
  attachments: Record<string, unknown>[];
  browser_name: string | null;
  browser_version: string | null;
  os_name: string | null;
  os_version: string | null;
};

export type IssueDetail = Issue & { sample_event: EventOut | null };

export type IssueUpdate = { status: "unresolved" | "resolved" | "ignored" };

File: backend/app/schemas/issues.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class IssueOut(BaseModel):
    id: UUID
    fingerprint: str
    title: str
    culprit: str | None
    level: str
    status: str
    first_seen_at: datetime
    last_seen_at: datetime
    event_count: int
    user_count: int

    model_config = {"from_attributes": True}


class IssueListResponse(BaseModel):
    items: list[IssueOut]
    next_cursor: str | None


class EventOut(BaseModel):
    id: UUID
    event_id: UUID
    occurred_at: datetime
    received_at: datetime
    environment: str
    release: str | None
    platform: str
    sdk_name: str | None
    sdk_version: str | None
    level: str
    message: str | None
    exception: dict[str, Any] | None
    request: dict[str, Any] | None
    user_: dict[str, Any] | None = Field(default=None, serialization_alias="user")
    tags: dict[str, Any]
    contexts: dict[str, Any]
    breadcrumbs: list[dict[str, Any]]
    attachments: list[dict[str, Any]]
    browser_name: str | None
    browser_version: str | None
    os_name: str | None
    os_version: str | None

    model_config = {"from_attributes": True, "populate_by_name": True}


class IssueDetailOut(IssueOut):
    sample_event: EventOut | None


class IssueUpdate(BaseModel):
    status: str = Field(..., pattern="^(unresolved|resolved|ignored)$")


class ProjectSummary(BaseModel):
    unresolved_count: int
    events_24h: int
    affected_users_24h: int
    unique_issues_24h: int


class FrequencyPoint(BaseModel):
    t: datetime
    count: int


class FrequencyResponse(BaseModel):
    range: str
    interval: str  # "hour" | "day"
    points: list[FrequencyPoint]
